lib/ExecutionEngine/ExecutionEngine2.cpp

```diff
@@ -0,0 +1,216 @@
+        // Don't delete context.
+        contextPtr.release();
+        runErr = std::move(err);
+        runPromise.set_value();
+      });
+
+  fut.wait();
+  EXIT_ON_ERR(std::move(runErr));
+}
+
+void ExecutionEngine::run(ExecutionContext &context) {
+  assert(compiledFunctions_.size() == 1 &&
+         "Expected exactly one compiled function.");
+  runInternal(context, *compiledFunctions_.begin());
+}
+
+void ExecutionEngine::run(ExecutionContext &context, llvm::StringRef name) {
+  runInternal(context, name);
+}
+
+void ExecutionEngine::run(PlaceholderBindings &bindings) {
+  assert(compiledFunctions_.size() == 1 &&
+         "Expected exactly one compiled function.");
+  std::unique_ptr<PlaceholderBindings> bindingsPtr(&bindings);
+  ExecutionContext context(std::move(bindingsPtr));
+  runInternal(context, *compiledFunctions_.begin());
+  // don't delete bindings
+  context.movePlaceholderBindings().release();
+}
+
+void ExecutionEngine::run(PlaceholderBindings &bindings, llvm::StringRef name) {
+  std::unique_ptr<PlaceholderBindings> bindingsPtr(&bindings);
+  ExecutionContext context(std::move(bindingsPtr));
+  runInternal(context, name);
+  // don't delete bindings
+  context.movePlaceholderBindings().release();
+}
+
+void glow::runBatch(ExecutionEngine &EE, PlaceholderBindings &bindings,
+                    size_t iterations, size_t &sampleCounter,
+                    llvm::ArrayRef<Placeholder *> ph,
+                    llvm::ArrayRef<Tensor *> inputs, llvm::StringRef name) {
+  // This is the size of one batch (the number of samples in the batch).
+  size_t batchSize = ph[0]->getType()->dims()[0];
+
+  assert(!inputs.empty() && "No inputs");
+  assert(inputs.size() == ph.size() &&
+         "The number of inputs does not match the number of placeholders");
+
+  // For each iteration in the batch:
+  for (size_t j = 0; j < iterations; j++) {
+
+    // Update the input placeholders.
+    for (int i = 0, e = ph.size(); i < e; i++) {
+      assert(ph[i] && "Invalid value");
+      auto *backingTensor = bindings.get(ph[i]);
+      assert(backingTensor && "Can't find the backing tensor");
+
+      auto dim = inputs[i]->dims();
+      assert(backingTensor->dims().drop_front() == dim.drop_front() &&
+             "Invalid slice size");
+      // Extract the n'th slice, that must be a tensor.
+      size_t slc = sampleCounter % dim[0];
+      // Pick up one slice from the input tensors, and load it into the
+      // corresponding network Placeholder.
+      backingTensor->copyConsecutiveSlices(inputs[i], slc);
+    }
+
+    // Run the network.
+    if (name == "") {
+      EE.run(bindings);
+    } else {
+      EE.run(bindings, name);
+    }
+    sampleCounter += batchSize;
+  }
+}
+
+void ExecutionEngine::compile(CompilationMode mode) {
+  CompilationContext cctx;
+  cctx.compMode = mode;
+  compile(cctx);
+}
+
+void ExecutionEngine::compile(CompilationContext &cctx) {
+  assert(module_.get() && "Compile has already been called.");
```

Comment:
So this call to compile() is terminal, in a way: once you call it you can't add any more functions or compile anything else until you setBackend() again? The old EE didn't work like that did it?

Reply:
It did not, this is because the HostManager compilation is module wide and destructive.